### ISSUE TITLE
Budget Allocation Percentage breakdown by channel

### ISSUE THREAD

**Issue by DongHarry-Kang** [google/lightweight_mmm#294]
Hi all - 

I am currently using budget allocator from lightweight, however I noticed the Before optimization budget is not correct. Could somebody let me know why this happened? for my analysis, n_time_periods = 10

![image](https://github.com/google/lightweight_mmm/assets/146807827/e1087df4-b66b-468b-8a34-f11d6cbfdc04)

<html xmlns:v="urn:schemas-microsoft-com:vml"
xmlns:o="urn:schemas-microsoft-com:office:office"
xmlns:x="urn:schemas-microsoft-com:office:excel"
xmlns="http://www.w3.org/TR/REC-html40">

<head>

<meta name=ProgId content=Excel.Sheet>
<meta name=Generator content="Microsoft Excel 15">
<link id=Main-File rel=Main-File
href="file:///C:/Users/DONGH~1.KAN/AppData/Local/Temp/msohtmlclip1/01/clip.htm">
<link rel=File-List
href="file:///C:/Users/DONGH~1.KAN/AppData/Local/Temp/msohtmlclip1/01/clip_filelist.xml">
<style>

</style>
</head>

<body link="#0563C1" vlink="#954F72">

Channel   0 | Channel 1 | Channel 2 | Channel 3 | Channel 4 | Channel 5 | Channel 6 | Channel 7 | Channel 8 | Channel 9
-- | -- | -- | -- | -- | -- | -- | -- | -- | --
36.7% | 33.5% | 15.1% | 5.8% | 3.8% | 1.2% | 2.5% | 1.2% | 0.1% | 0.0%

</body>

</html>

**Comment by entzyeung** [comment]
Dear LMMM Team,

I want to express my sincere appreciation for the hard work you've dedicated to creating this remarkable model. It's truly impressive and has proven to be user-friendly and easy to install, unlike my previous attempt with the FB Robyn model, which I eventually abandoned due to its complexity. Discovering your model felt like a ray of sunshine after a cloudy day.

To cut to the chase, I've encountered a similar issue to that of @DongHarry-Kang. I've spent considerable time experimenting with various datasets and fine-tuning for two weeks. However, I consistently obtain similar results when calling the `plot_pre_post_budget_allocation_comparison` function.

I've included some sample outputs below:

![Output 1](https://github.com/google/lightweight_mmm/assets/60934147/398725ef-c337-4f6c-9599-0dd14218ca54)

![Output 2](https://github.com/google/lightweight_mmm/assets/60934147/eb904362-c22d-49bd-9406-b1f8a76e1e64)

I've thoroughly examined data quality using `check_data_quality()`, assessed variances with `highlight_variances()`, reviewed spend fractions with `highlight_low_spend_fractions()`, and analyzed variance inflation factors with `highlight_high_vif_values()`. The datasets appear to be in good shape and suitable for analysis. I even trained the best model with parameters obtained from Gridsearches. However, when it comes to the critical aspect of budget allocation, it seems there may be an issue.

Regardless of how I adjust the budget and cost parameters, the model consistently produces the same outcome: maintaining the same channel ratios while suggesting a higher budget allocation.

Allow me to illustrate this with an example.
<img width="705" alt="Screenshot 2024-01-07 at 22 55 23" src="https://github.com/google/lightweight_mmm/assets/60934147/77b11f6d-173e-4a19-9d53-ac24aba14dec">

As you can see from the investigation, the model did suggest me different and higher budget while maintaining the same channel ratios. I am so confused that why the model not reallocating the budgets as the documentation said? and keep increasing our budget instead?
is there any params I missed? if not, is there any way to fix it?

Here is my syntax when calling the optimizatin function:
# Run optimization with the parameters of choice.
```
solution, kpi_without_optim, previous_media_allocation = optimize_media.find_optimal_budgets(
            n_time_periods= n_time_periods,
            media_mix_model= mmm,
            extra_features= extra_features_test[:n_time_periods],# it has been scaled.
            budget= budget,
            prices= prices,
            media_scaler = media_scaler,
            target_scaler = target_scaler,
            )
```

# Plot out pre post optimization budget allocation and predicted target variable comparison.
```
plot.plot_pre_post_budget_allocation_comparison(media_mix_model=mmm, 
                                                kpi_with_optim=solution['fun'], 
                                                kpi_without_optim=kpi_without_optim,
                                                optimal_buget_allocation=optimal_buget_allocation, 
                                                previous_budget_allocation=previous_budget_allocation, 
                                                figure_size=(10,10))
```
Thank you in advance.

**Comment by shubhamgupta568** [comment]
@entzyeung can you please share your code ? I too faced similar issue and I am able to resolve it.

**Comment by entzyeung** [comment]
@shubhamgupta568 thank you for dropping by, here is the link to the [notebook](https://drive.google.com/file/d/1NJRoyaLD3mkyCpRKESzoCjYknAs_7HbV/view?usp=sharing).
please let me know why was that and which pieces of code should be corrected. Appreciated!

**Comment by shubhamgupta568** [comment]
@entzyeung I can see only 2 problems. How you defined your prices and budget. I am using below code to find price and budget. Please try this and check if this fix your problem.
Note, here costs and media_data are unscaled values.

prices = costs / media_data.sum(axis=0)
budget = jnp.sum(jnp.dot(prices, media_data.mean(axis=0)))* n_time_periods

**Comment by entzyeung** [comment]
Thank you @shubhamgupta568 for the solution, i am running the code again now. The first run gave me the same distribution, so i am running it again. WIll keep you posted. What caused the problem? is it the dtype of the media_data?

**Comment by entzyeung** [comment]
Hello @shubhamgupta568 , I have tried many times, and have made sure both costs, and media_data are unscaled jnp array. Here are the results:
<img width="865" alt="Screenshot 2024-01-11 at 21 17 49" src="https://github.com/google/lightweight_mmm/assets/60934147/45a5e35d-c161-4e6b-a562-8741640db17f">
<img width="867" alt="Screenshot 2024-01-11 at 21 18 07" src="https://github.com/google/lightweight_mmm/assets/60934147/d6306845-2e56-4314-a3b7-1ba42775c1ba">

As you can see, the budget allocation percentages are the same, just the total budget is inflated a bit. I was thinking if it was the data type problem. But I have ensured they are unscaled jnp array. So i really have no clue what is happening here.

**Comment by shubhamgupta568** [comment]
I think the reason you might be getting same distribution can be the upper and lower bound percentages. Can you try changing that too ? May be something like upper bound 2 and lower bound 0.5

**Comment by entzyeung** [comment]
I haven't yet set it directly. I was just using default. In the predict
function?

**Comment by shubhamgupta568** [comment]
@entzyeung  Its part of find_optimal_budget function. 
![image](https://github.com/google/lightweight_mmm/assets/46197367/35d745fb-d086-4997-b577-86b4dd861443)

**Comment by entzyeung** [comment]
code:
```
solution, kpi_without_optim, previous_media_allocation = optimize_media.find_optimal_budgets(
    n_time_periods= n_time_periods,
    media_mix_model= mmm,
    extra_features= extra_features_test[:n_time_periods],#### transformed, and the forecasted period [:n_time_periods]
    budget= budget,
    prices= prices,
    media_scaler = media_scaler,
    target_scaler = target_scaler,

    seed= SEED,
    bounds_upper_pct = 0.05, 
    bounds_lower_pct = 0.05
    )
```
I tried multiple sets of upper and lower pct, all of them gave me the same percentage of distribution, but if i set the upper and lower smaller than 0.2, there would be a height differences between bars, while remaining the same pct. If i set the upper and lower bigger than 0.2, the height between bars would be exactly the same. 

The graph below is bounds pct = 0.3
<img width="701" alt="0 3" src="https://github.com/google/lightweight_mmm/assets/60934147/6c3712b1-9fa0-424d-aa15-cb3ac2e1358a">

The graph below is bounds pct = 0.05
<img width="702" alt="0 05" src="https://github.com/google/lightweight_mmm/assets/60934147/2a539de5-38a7-4746-aee5-81cb7a35ca50">
